### PR TITLE
kafka(ticdc): Hide async producer sarama

### LINKED ISSUE #8047: kafka async producer hide sarama

### PR DESCRIPTION
**Pull request by sdojjy**
### What problem does this PR solve?

Issue Number: close #8047

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: pkg/sink/kafka/client.go [hunk 1]
```diff
@@ -72,21 +76,16 @@ type AsyncProducer interface {
...
-
-	// Errors is the error output channel back to the user. You MUST read from this
-	// channel or the Producer will deadlock when the channel is full. Alternatively,
-	// you can set Producer.Return.Errors in your config to false, which prevents
-	// errors to be returned.
-	Errors() <-chan *sarama.ProducerError
+	AsyncSend(context.Context, string, int32, []byte, []byte,
+		chan struct{}, func()) error
+
+	AsyncCallbackRun(ctx context.Context,
```
Comment (3AceShowHand) [inline pkg/sink/kafka/client.go:84]: please add comment about this method.

File: pkg/sink/kafka/client.go [hunk 2]
```diff
@@ -72,21 +76,16 @@ type AsyncProducer interface {
...
-	// Producer will deadlock. It is suggested that you send and read messages
-	// together in a single select statement.
-	Successes() <-chan *sarama.ProducerMessage
-
-	// Errors is the error output channel back to the user. You MUST read from this
-	// channel or the Producer will deadlock when the channel is full. Alternatively,
-	// you can set Producer.Return.Errors in your config to false, which prevents
-	// errors to be returned.
-	Errors() <-chan *sarama.ProducerError
+	AsyncSend(context.Context, string, int32, []byte, []byte,
```
Comment (3AceShowHand) [inline pkg/sink/kafka/client.go:81]: please also add variable names to make the method easy to understand.

File: pkg/sink/kafka/client.go [hunk 3]
```diff
@@ -174,18 +173,70 @@ func (p *saramaAsyncProducer) Close() error {
...
-// wish to send.
-func (p *saramaAsyncProducer) Input() chan<- *sarama.ProducerMessage {
-	return p.producer.Input()
-}
-
-func (p *saramaAsyncProducer) Successes() <-chan *sarama.ProducerMessage {
-	return p.producer.Successes()
+func (p *saramaAsyncProducer) AsyncCallbackRun(ctx context.Context,
+	changefeedID model.ChangeFeedID,
+	closedChan chan struct{},
```
Comment (3AceShowHand) [inline pkg/sink/kafka/client.go:178]: for `changefeed`, `closeChan` and `failpointCh`, can we put them into the `saramaAsyncProducer` struct as fields, to make the signature simple.

File: pkg/sink/kafka/client.go [hunk 4]
```diff
@@ -174,18 +173,70 @@ func (p *saramaAsyncProducer) Close() error {
...
-func (p *saramaAsyncProducer) Errors() <-chan *sarama.ProducerError {
-	return p.producer.Errors()
+// AsyncSend is the input channel for the user to write messages to that they
+// wish to send.
+func (p *saramaAsyncProducer) AsyncSend(ctx context.Context,
+	topic string,
+	partition int32,
+	key []byte,
+	value []byte,
+	closedChan chan struct{},
```
Comment (3AceShowHand) [inline pkg/sink/kafka/client.go:222]: ditto, put `closedChan` into `p`.

File: pkg/sink/kafka/client.go (hunk 1 above)
Comment (sdojjy) [reply pkg/sink/kafka/client.go:84]: done

File: pkg/sink/kafka/client.go (hunk 3 above)
Comment (sdojjy) [reply pkg/sink/kafka/client.go]: done

File: pkg/sink/kafka/client.go (hunk 4 above)
Comment (sdojjy) [reply pkg/sink/kafka/client.go]: done

**Review by asddongmen — APPROVED** [review]

**Pull request merged**